### PR TITLE
chore: hive schema inference

### PR DESCRIPTION
**Pull request by Clark0**
### Changelog category :
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
experimental hive table supports schema inference & refactor hive code
...

### Documentation entry for user-facing changes
Support create hive table like `CREATE TABLE hive.tx ENGINE = CnchHive('thrift://localhost:9183', 'par_db', 'par_tbl')
` and column descriptions can be deduced from the hive metastore

### CODE REVIEW

**Comment by Clark0** [comment]
relates to #315

File: src/Storages/StorageFactory.h [hunk 1]
```diff
@@ -88,6 +88,7 @@ class StorageFactory : private boost::noncopyable, public IHints<1, StorageFacto
         bool supports_deduplication = false;
         /// See also IStorage::supportsParallelInsert()
         bool supports_parallel_insert = false;
+        bool supports_schema_inference;
```
Comment (wordhardqi) [inline src/Storages/StorageFactory.h:91]: follow 上面设置为 false?

File: src/Storages/Hive/HiveSchemaConverter.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+{
+    size_t pos = hive_type_name.find('<');
+    if (pos != String::npos)
+    {
+        size_t right_pos = hive_type_name.find('>');
+        size_t len = right_pos - pos - 1;
+        return {hive_type_name.substr(0, pos), hive_type_name.substr(pos + 1, len)};
+    }
+
+    pos = hive_type_name.find('(');
```
Comment (wordhardqi) [inline src/Storages/Hive/HiveSchemaConverter.cpp:41]: 啥时候回用到()

File: src/Storages/Hive/HiveSchemaConverter.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+DataTypePtr HiveSchemaConverter::hiveTypeToCHType(const String & hive_type, bool make_columns_nullable)
+{
+    static const std::unordered_map<String, std::shared_ptr<IDataType>> base_type_mapping = {
+        {"tinyint", std::make_shared<DataTypeInt8>()},
+        {"smallint", std::make_shared<DataTypeInt16>()},
+        {"bigint", std::make_shared<DataTypeInt64>()},
+        {"int", std::make_shared<DataTypeInt32>()},
+        {"integer", std::make_shared<DataTypeInt32>()},
+        {"float", std::make_shared<DataTypeFloat32>()},
+        {"decimal", std::make_shared<DataTypeFloat32>()},
```
Comment (wordhardqi) [inline src/Storages/Hive/HiveSchemaConverter.cpp:60]: 这个不对, clickhouse 也有decimal. 然后decimal 是 Decimal(10,0) 这种, parse的有点问题.

File: src/Storages/Hive/HiveSchemaConverter.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    else if (type_keyword == "map")
+    {
+        Strings res;
+        boost::split(res, inner, boost::is_any_of(","), boost::token_compress_on);
+        auto key_type = hiveTypeToCHType(res.at(0), false);
+        auto value_type = hiveTypeToCHType(res.at(1), false);
+        data_type = std::make_shared<DataTypeByteMap>(key_type, value_type);
+    }
+    else if (type_keyword == "struct")
+    {
```
Comment (wordhardqi) [inline src/Storages/Hive/HiveSchemaConverter.cpp:101]: 要不要直接丢个异常,说不支持.

File: src/Storages/Hive/HiveSchemaConverter.cpp [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+{
+}
+
+StorageInMemoryMetadata HiveSchemaConverter::convert() const
+{
+    ColumnsDescription columns;
+    auto addColumn = [&](const Apache::Hadoop::Hive::FieldSchema & hive_field) {
+        // bool make_columns_nullable = getContext()->getSettingsRef().data_type_default_nullable;
+        bool make_columns_nullable = true;
+        DataTypePtr ch_type = hiveTypeToCHType(hive_field.type, make_columns_nullable);
```
Comment (wordhardqi) [inline src/Storages/Hive/HiveSchemaConverter.cpp:123]: Hive column的comment 要不要拿过来.

Comment (wordhardqi) [reply src/Storages/Hive/HiveSchemaConverter.cpp:123]: 这里也不对, Hive 可以指定某列Not Null

**Review by wordhardqi — COMMENTED** [review]
plz check the decimal type and not-null constraint.

File: src/Storages/Hive/HiveSchemaConverter.cpp (hunk 4 above)
Comment (Clark0) [reply src/Storages/Hive/HiveSchemaConverter.cpp:101]: We first omit the column with unsupported type, and add support in the future

**Review by wordhardqi — APPROVED** [review]

**Review by wordhardqi — APPROVED** [review]

**Review by wordhardqi — APPROVED** [review]

File: src/Storages/Hive/HiveSchemaConverter.cpp (hunk 3 above)
Comment (Clark0) [reply src/Storages/Hive/HiveSchemaConverter.cpp]: fixed

**Review by wordhardqi — APPROVED** [review]

**Pull request merged**